### DISCUSSION TITLE
`[Management.Automation.WildcardPattern]::Escape("$Path")` doesn't escape all characters. Should it? If not, why?

### DISCUSSION THREAD

**Discussion by roke-julian-lockhart** [PowerShell/PowerShell#24017] [Q&A] [answered]
Today, I had the same question as [this on StackOverflow](https://stackoverflow.com/revisions/46037095/4#:~:text=Is%20there%20a%20function%20in%20PowerShell%20for%20escaping%20characters%20in%20paths?):

> Is there a function in PowerShell for escaping characters in paths?

[The accepted answer to it](https://stackoverflow.com/revisions/46037546/2#:~:text=you%20can%20use:-,%5BManagement.Automation.WildcardPattern%5D::Escape('test%5B1%5D.txt'),-Returns:) appears to function on the provided example:

<blockQuote>

There is the following method you can use:

```PS1
[Management.Automation.WildcardPattern]::Escape('test[1].txt')
```

Returns:

```log
test`[1`].txt
```

</blockQuote>

However, if I craft a path of a myriad of UTF-8 characters, like <kbd>tab</kbd>s and backticks (<kbd>`</kbd>s) it appears to fail easily:

1.	```PS1
	Get-Content "$([Management.Automation.WildcardPattern]::Escape('/home/RokeJulianLockhart/Software/Git/GitLab.com/RokeJulianLockhart/sg5n9q/  [-]	`{title: "Non-Metadata",	id: "s1gvg2"}`{.JSON5}.dir/  [-]	`{title: "Documents",	id: "s4aaax"}`{.JSON5}.dir/  [-]	`{title: "Stories",	id: "s329sp"}`{.JSON5}.dir/  [-]	`{title: "Content",	id: "s20cw5"}`{.JSON5}.txt.MD'))"
	```
	
1.	<blockQuote>

	```log
	PS /home/RokeJulianLockhart> Get-Content "$([Management.Automation.WildcardPattern]::Escape('/home/RokeJulianLockhart/Software/Git/GitLab.com/RokeJulianLockhart/sg5n9q/  [-]`{title: "Non-Metadata",id: "s1gvg2"}`{.JSON5}.dir/  [-]`{title: "Documents",id: "s4aaax"}`{.JSON5}.dir/  [-]`{title: "Stories",id: "s329sp"}`{.JSON5}.dir/  [-]`{title: "Content",id: "s20cw5"}`{.JSON5}.txt.MD'))"
	Get-Content: Cannot find path '/home/RokeJulianLockhart/Software/Git/GitLab.com/RokeJulianLockhart/sg5n9q/  [-]`{title: "Non-Metadata",id: "s1gvg2"}`{.JSON5}.dir/  [-]`{title: "Documents",id: "s4aaax"}`{.JSON5}.dir/  [-]`{title: "Stories",id: "s329sp"}`{.JSON5}.dir/  [-]`{title: "Content",id: "s20cw5"}`{.JSON5}.txt.MD' because it does not exist.
	PS /home/RokeJulianLockhart>
	```

	</blockQuote>

I'm using BTRFS on a non-AOSP Linux-based OS (`cpe:/o:fedoraproject:fedora:40`) so I can create paths with such characters.

Is this behaviour expected? If so, why? If not, please confirm so that I can file a bug.

**Comment by rhubarb-geek-nz** [comment]
I suggest you show a concise reproducible problem, your example looks like trying to put file content into a filename (which is not expected behaviour). Narrow it down to the simplest examples that demonstrate the problem, not the most complex.

I notice that you have forward slashes within the filename, are these are subdirectories?  eg where is says `{.JSON5}.dir/ ?

Does using `GetContent -LiteralPath` make any difference?

**Reply by roke-julian-lockhart** [reply]
> https://github.com/PowerShell/PowerShell/discussions/24017#discussioncomment-9977852

@rhubarb-geek-nz, apologies for being lazy. Undermentioned is an example which should demonstrate the issue more concisely, since it solely includes a tab and Latin ASCII:

1.	1.	```PS1
		ls "$PWD" --time-style='full-iso' --inode -l --group-directories-first 
		```
	
	1.	<blockQuote>
		
		```log
		PS /home/RokeJulianLockhart/test> ls "$PWD" --time-style='full-iso' --inode -l --group-directories-first 
		total 0
		20817476 drwxr-xr-x. 1 RokeJulianLockhart RokeJulianLockhart 0 2024-07-07 15:48:36.021648426 +0100 ''$'\t''tab'
		20686815 drwxr-xr-x. 1 RokeJulianLockhart RokeJulianLockhart 0 2024-07-05 17:42:00.986610195 +0100 '  [-]'$'\t''`{title: "Non-Metadata",'$'\t''id: "s4a4ji"}`{.JSON5}.dir'
		```
		
		</blockQuote>
	
1.	1.	```PS1
		Get-Content -LiteralPath "$([Management.Automation.WildcardPattern]::Escape('/home/RokeJulianLockhart/test/tab'))"
		```
	
	1.	<blockQuote>
		
		```
		PS /home/RokeJulianLockhart/test> Get-Content -LiteralPath "$([Management.Automation.WildcardPattern]::Escape('/home/RokeJulianLockhart/test/tab'))"
		Get-Content: Cannot find path '/home/RokeJulianLockhart/test/tab' because it does not exist.
		PS /home/RokeJulianLockhart/test>
		```
		
		</blockQuote>

**Comment by rhubarb-geek-nz** [comment, marked as answer]
It is worth seeing what [System.Management.Automation.WildcardPattern.Escape](https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.wildcardpattern.escape?view=powershellsdk-7.4.0) actually does.

```
char c = (char)0;
char[] ac = new char[1];

do
{
    ac[0]= c;
    string s = new string(ac);

    string r = System.Management.Automation.WildcardPattern.Escape(s);

    if (!r.Equals(s))
    {
        System.Console.WriteLine($"{s} -> {r}");
    }

} while (0 != ++c);
```

Gives

```
* -> `*
? -> `?
[ -> `[
] -> `]
```

So out of the entire UNICODE character set it only escapes '*','?','[' and ']', which we can presume are the special characters the documentation refers to, ie, the characters used by PowerShell in its wildcard expansions.

Given that it is not wrapped as a cmdlet, as far as I can tell and PowerShell Core >=7.3 has a special mistrust of anyone using method invocation, I can only presume it is not for general use.

**Reply by roke-julian-lockhart** [reply]
> https://github.com/PowerShell/PowerShell/discussions/24017#discussioncomment-9979450

@rhubarb-geek-nz, that would indeed explain it. Thanks. It appears to be solely designed for use in Windows. Perhaps it'd be worth requesting it escape a little more at the DotNet repository, unless this is an X/Y problem because there's something better designed for this purpose.

**Reply by jborean93** [reply]
> It appears to be solely designed for use in Windows

It's designed to escape wildcard patterns which is `*`, `?`, `[`, and `]`. All other characters can be used as is with the `-Path` parameter. If there is a particular character that fails then it's probably worth investigating why that is and seeing if it can be fixed. It isn't the tab character though, it works just fine in my example

```powershell
PS /Users/jborean> python -c "import os; os.mkdir('$([System.IO.Path]::GetTempPath())/test`ttab')"
PS /Users/jborean> Get-Item "$([System.IO.Path]::GetTempPath())/test`ttab"

    Directory: /var/folders/7y/bc_81g817_q985d7h7m5gsnc0000gn/T

UnixMode         User Group         LastWriteTime         Size Name
--------         ---- -----         -------------         ---- ----
drwxr-xr-x    jborean staff        8/7/2024 04:55           64 test     tab

PS /Users/jborean> Get-Item ([WildcardPattern]::Escape("$([System.IO.Path]::GetTempPath())/test`ttab"))

    Directory: /var/folders/7y/bc_81g817_q985d7h7m5gsnc0000gn/T

UnixMode         User Group         LastWriteTime         Size Name
--------         ---- -----         -------------         ---- ----
drwxr-xr-x    jborean staff        8/7/2024 04:55           64 test     tab
```

**Reply by roke-julian-lockhart** [reply]
> https://github.com/PowerShell/PowerShell/discussions/24017#discussioncomment-9980758

@jborean93,

```
Get-Item ([WildcardPattern]::Escape("test/`ttab"))
```

works for me too. Thanks.

However, is there no method which allows me to merely paste the path verbatim, and have it automatically escaped so that it's parsed correctly?

**Reply by jborean93** [reply]
If you mean a path from the `ls` output then I'm not aware of any builtin way. You need to use PowerShell's escaping mechanism which uses backticks rather than backslashes or build your own escaping mechanism to do so.

**Reply by roke-julian-lockhart** [reply]
> https://github.com/PowerShell/PowerShell/discussions/24017#discussioncomment-9980855

@jborean93, I'll build something, then. Thanks.

**Comment by roke-julian-lockhart** [comment]
I think I've found the cause of my woes. I've reported it at https://github.com/PowerShell/PowerShell/issues/24037#issue-2400529452. It would explain why all the code examples worked.
